### PR TITLE
rustbuild changes to build LLVM

### PR DESCRIPTION
**Pull request by Zoxc**
This fixes issues with --llvm-root in msys2.

### CODE REVIEW

File: src/bootstrap/build/util.rs [hunk 1]
```diff
@@ -16,6 +16,26 @@ use std::process::Command;
 use bootstrap::{dylib_path, dylib_path_var};
 use filetime::FileTime;
 
+#[cfg(not(windows))]
+pub fn build_path(path: &str) -> PathBuf {
+    PathBuf::from(path)
+}
+
+#[cfg(windows)]
+pub fn build_path(path: &str) -> PathBuf {
```
Comment (alexcrichton) [inline src/bootstrap/build/util.rs]: Hm this seems very suspicious about how this would be called, how are MSYS paths leaking into the configuration here? Are the makefiles encoding MSYS paths?

Also, this can probably be done in Rust like [`sanitize_sh`](https://github.com/rust-lang/rust/blob/70b8b9487a94c5b9b1401007b040c584d9349582/src/bootstrap/build/dist.rs#L273-L290) rather than forking out to `cygpath`, right?

File: src/librustc_llvm/build.rs [hunk 2]
```diff
@@ -35,7 +35,7 @@ fn main() {
         }
         PathBuf::from("llvm-config")
     });
-
+    llvm_config.set_extension(build_exe());
```
Comment (alexcrichton) [inline src/librustc_llvm/build.rs]: Could this be `std::os::consts::EXE_SUFFIX`?

File: src/librustc_llvm/build.rs [hunk 3]
```diff
@@ -160,6 +161,9 @@ fn main() {
             println!("cargo:rustc-link-search=native={}", &lib[2..]);
         }
     }
+    // Add <llvm-root>/bin to the search path, DLLs will be placed here
+    // llvm-config is bugged and doesn't doesn't indicate this on Windows
+    println!("cargo:rustc-link-search=native={}", llvm_config.join("..").to_str().unwrap());
```
Comment (alexcrichton) [inline src/librustc_llvm/build.rs]: Could this instead do `llvm_config.pop()` and print that?

File: src/bootstrap/build/native.rs [hunk 4]
```diff
@@ -146,9 +146,12 @@ pub fn compiler_rt(build: &Build, target: &str) {
...
     t!(fs::create_dir_all(&dst));
-    let build_llvm_config = build.llvm_out(&build.config.build)
-                                 .join("bin")
-                                 .join(exe("llvm-config", &build.config.build));
+
+    let build_llvm_config = build.config.target_config.get(target).and_then(|config| {
+      config.llvm_config.clone()
+    }).unwrap_or(build.llvm_out(&build.config.build)
+                      .join("bin")
+                      .join("llvm-config"));
```
Comment (alexcrichton) [inline src/bootstrap/build/native.rs]: Could you reformat this to be more like:

```rust
let default = ...;
let value = thing_with_option.unwarp_or(default);
```

File: src/bootstrap/build/util.rs (hunk 1 above)
Comment (Zoxc) [inline src/bootstrap/build/util.rs]: The `configure` script produces MSYS path, so we need to convert them when importing `config.mk`.

I'm not sure what the logic is to convert to Windows paths, so I'd rather avoid writing that.

File: src/bootstrap/bootstrap.py [hunk 5]
```diff
@@ -180,6 +180,7 @@ def parse_nightly_dates(self):
 
     def build_bootstrap(self):
         env = os.environ.copy()
+        env["CARGO_HOME"] = os.path.join(self.build_dir, "cargo")
```
Comment (alexcrichton) [inline src/bootstrap/bootstrap.py]: How come this was added? In theory this should pick up the same `CARGO_HOME` that it inherits?

File: src/bootstrap/build/mod.rs [hunk 6]
```diff
@@ -501,6 +501,18 @@ impl Build {
...
+        let external_llvm = self.config.target_config.get(target).and_then(|config| {
+            config.llvm_config.clone()
+        });
+        if let Some(mut llvm_config) = external_llvm {
+            llvm_config.pop();
+            llvm_config.pop();
+            // Search in <llvm-root>/bin for shared libraries on Windows
+            base.push(format!("-L{}", llvm_config.join("bin").to_str().unwrap()));
+            // Search in <llvm-root>/lib elsewhere
+            base.push(format!("-L{}", llvm_config.join("lib").to_str().unwrap()));
```
Comment (alexcrichton) [inline src/bootstrap/build/mod.rs]: Hm this is something that I would very much to avoid like the plague with rustbuild, these sorts of flags should all be handled by Cargo, not our own build system. How come these needed to be added?

File: src/bootstrap/build/util.rs (hunk 1 above)
Comment (alexcrichton) [inline src/bootstrap/build/util.rs]: The logic should be pretty simple, right? `/a/foo` translates to `A:\foo`?

File: src/bootstrap/build/mod.rs (hunk 6 above)
Comment (Zoxc) [inline src/bootstrap/build/mod.rs]: The problem is that `rustc_llvm` is in the sysroot, so its dependencies are not handled by cargo.

See https://github.com/rust-lang/rust/issues/30970

File: src/bootstrap/build/util.rs (hunk 1 above)
Comment (Zoxc) [inline src/bootstrap/build/util.rs]: That's only the Windows paths in msys form. There are also things like `/home` and even `/h/path` where `/h` as a directory and not a drive.

File: src/bootstrap/bootstrap.py (hunk 5 above)
Comment (Zoxc) [inline src/bootstrap/bootstrap.py]: This was added so rustbuild doesn't write to directories outside the build directory and the rust source.

Comment (alexcrichton) [inline src/bootstrap/bootstrap.py]: Let's leave this out for now as it seems unrelated to this change, and this can always be set externally or we can add a configuration option for it.

File: src/bootstrap/build/mod.rs (hunk 6 above)
Comment (alexcrichton) [inline src/bootstrap/build/mod.rs]: I would personally like to not land this patch if we're adding linker flags like this, this is not where they're supposed to go. Essentially 100% of the logic of dealing with LLVM should happen as part of `src/librustc_llvm/build.rs`.

Where is the build failing if these aren't specified? This basically means that all external users which do that kind of build will *also* fail which means that we shouldn't require these flags to compile.

File: src/bootstrap/build/util.rs (hunk 1 above)
Comment (alexcrichton) [inline src/bootstrap/build/util.rs]: Right so this can cause problems, but this happens on every startup of the build system, so it would be best to spawn fewer processes and just catch up on edge cases over time. The vast majority of these paths will just be `/c/foo` or `/other` that we need to fix (which are all trivial to fix)

Comment (alexcrichton) [inline src/bootstrap/build/util.rs]: Or actually, even better, we should change the `configure` script to not emit MSYS paths, but rather use `cygpath` in there.

Comment (Zoxc) [inline src/bootstrap/build/util.rs]: I don't want to touch that script though.

File: src/bootstrap/build/mod.rs (hunk 6 above)
Comment (Zoxc) [inline src/bootstrap/build/mod.rs]: rustbook needs it to build, but that's not required when disabling docs (and generating docs is broken anyway).

A proper solution would be to hardcode the LLVM linking path into either rustc_llvm's binaries or into the rustc executable.

File: src/bootstrap/build/sanity.rs [hunk 7]
```diff
@@ -78,6 +79,24 @@ pub fn check(build: &mut Build) {
             panic!("the iOS target is only supported on OSX");
         }
 
+        if cfg!(windows) {
```
Comment (alexcrichton) [inline src/bootstrap/build/sanity.rs]: Could you add a comment to this explaining that this is just ensuring that we can run binaries like `llvm-config` and the compiler by ensuring that the dynamically linked libraries are in `PATH`?

Comment (alexcrichton) [inline src/bootstrap/build/sanity.rs]: This also seems relevant for all platforms (not just Windows), so perhaps this could create a variable like `ldpath` to handle the Unix/OSX cases?

File: src/bootstrap/build/util.rs (hunk 1 above)
Comment (alexcrichton) [inline src/bootstrap/build/util.rs]: I don't think we can stomach all of these process spawns on *every invocation of the build system*, so this needs to be handled in some method.

File: src/bootstrap/build/sanity.rs (hunk 7 above)
Comment (Zoxc) [inline src/bootstrap/build/sanity.rs]: This isn't needed on Linux atleast, since it links using full paths and not just the filename,

Comment (alexcrichton) [inline src/bootstrap/build/sanity.rs]: Isn't this needed for running `llvm-config` at all? Or running the compiler if it's dynamically linked? I may be misunderstanding the purpose for this though?

Comment (Zoxc) [inline src/bootstrap/build/sanity.rs]: It is not need for `llvm-config` on Windows, since it's in the directory with the DLLs. It is needed for executables which are not in the LLVM's binary directory which link dynamically to LLVM (like rustc).

Comment (alexcrichton) [inline src/bootstrap/build/sanity.rs]: Right, but that's not a Windows specific limitation, right? If rustc ever dynamically links to LLVM (on any platform) we need the library directory to be in the dynamic library path?

It looks like the difference is just that the dynamic library directory on Windows happens to be the same as the binaries directory?

Comment (Zoxc) [inline src/bootstrap/build/sanity.rs]: It is Windows specific, as you link with filenames on Windows and not paths.

Comment (alexcrichton) [inline src/bootstrap/build/sanity.rs]: Sorry I think I'm not being clear enough. So if you're liking with an external LLVM that is itself dynamically linked, then the only hope we have of working correctly is to ensure that the directory with that LLVM dynamic library is in the runtime linker lookup path. On Windows this just happens to be `PATH` but on Linux/OSX this id `LD_LIBRARY_PATH` and `DYLD_LIBRARY_PATH`.

Which is to say, if we link to a dynamically linked LLVM on *any* platform, we need to run this logic to ensure that the relevant directory is in the relevant environment variable.

**Comment by alexcrichton** [comment]
Closing due to inactivity, but feel free to resubmit with a rebase!
